source/shared/cpp/ObjectModel/TextBlock.cpp

```diff
@@ -165,3 +174,144 @@ void TextBlock::SetHorizontalAlignment(const HorizontalAlignment value)
+    bool isDate = false; 
+    int  formatStyle = 0;
+    enum MatchIndex
+    {
+        IsDate = 2,
+        Year = 4,
+        Month,
+        Day,
+        Hour,
+        Min,
+        Sec,
+        TimeZone = 12,
+        TZHr,
+        TZMn,
+        Format,
+        Style,
+    };
+    int factor = 0, hours = 0, minutes = 0;
+    struct tm parsedTm = { 0 };
+    int *addrs[] = {&parsedTm.tm_year, &parsedTm.tm_mon,
+        &parsedTm.tm_mday, &parsedTm.tm_hour, &parsedTm.tm_min,
+        &parsedTm.tm_sec, &hours, &minutes};
+    std::vector<int> indexer = {Year, Month, Day, Hour, Min, Sec, TZHr, TZMn};
+
+    while (std::regex_search(text, matches, pattern))
+    {
+        // Date is matched
+        isDate = matches[IsDate].matched;
+
+        if(matches[Style].matched)
+        {
+            // match for long/short/compact
+            formatStyle = matches[Format].str()[2];
+        }
+
+        parsedostr << matches.prefix().str();
+
+        if(!isDate && formatStyle)
+        {
+            parsedostr << matches[0].str();
+            text = matches.suffix().str();
+            continue;
+        }
+
+        for (unsigned int idx = 0; idx < indexer.size(); idx++)
+        {
+            if (matches[indexer[idx]].matched)
+            {
+                // get indexes for time attributes to index into conrresponding matches
+                // and covert it to string
+                *addrs[idx] = stoi(matches[indexer[idx]]);
+            }
+        }
+
+        // maches offset sign, 
+        // Z == UTC, 
+        // + == time added from UTC
+        // - == time subtracted from UTC
+        if(matches[TimeZone].matched)
+        {
+            
+            char zone = matches[TimeZone].str()[0];
+            factor = (zone == '+')? -1 : 1;
+        }
+
+        // converts to seconds
+        hours *= 3600;
+        minutes *= 60;
+
+        // time zone offset calculation
+        offset = (hours + minutes) * factor;
+
+
+        // measured from year 1900
+        parsedTm.tm_year -= 1900;
+        parsedTm.tm_mon  -= 1;
+
+        time_t utc;
+        // converts to ticks in UTC
+        utc = mktime(&parsedTm);
+        if (utc == -1)
+        {
+            parsedostr << matches[0];
+        }
+
+        char tzOffsetBuff[6] = { 0 };
+        // gets local time zone offset
+        strftime(tzOffsetBuff, 6, "%z", &parsedTm);
+        std::string localTimeZoneOffsetStr(tzOffsetBuff);
+        int nTzOffset = std::stoi(localTimeZoneOffsetStr);
+        offset += ((nTzOffset / 100) * 3600 + (nTzOffset % 100) * 60);
+        // add offset to utc
+        utc += offset;
+        struct tm result = {0};
+
+        // converts to local time from utc
+        if (!LOCALTIME(&result, &utc))
+        {
+            // localtime_s double counts daylight saving time
+            if (result.tm_isdst == 1)
+                result.tm_hour -= 1;
+
+            if (isDate)
+            {
+                switch(formatStyle)
```

Comment:
What would be the effect here if we add a new style that starts with one of these letters in the future?

Reply:
that's something we have to deal with later. i don't want to make it more complex than necessary.

---
In reply to: [150952262](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150952262) [](ancestors = 150952262)

Reply:
Seems like if we're aware that it's going to be an issue, it should be addressed. Is it too big of a change to get the whole format string rather than the first character?

---
In reply to: [150965213](https://github.com/Microsoft/AdaptiveCards/pull/871#discussion_r150965213) [](ancestors = 150965213,150952262)